src/main/java/org/wildfly/security/credential/source/OAuth2CredentialSource.java

```diff
@@ -0,0 +1,277 @@
+package org.wildfly.security.credential.source;
+
+import static org.wildfly.common.Assert.checkNotNullParam;
+import static org.wildfly.security._private.ElytronMessages.log;
+
+import javax.json.Json;
+import javax.json.JsonObject;
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.security.AccessController;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.wildfly.security.auth.SupportLevel;
+import org.wildfly.security.auth.client.AuthenticationContext;
+import org.wildfly.security.auth.client.AuthenticationContextConfigurationClient;
+import org.wildfly.security.credential.BearerTokenCredential;
+import org.wildfly.security.credential.Credential;
+import org.wildfly.security.util.ByteStringBuilder;
+
+/**
+ * A {@link CredentialSource} capable of authenticating against a OAuth2 compliant authorization server and obtaining
+ * access tokens in form of a {@link BearerTokenCredential}.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class OAuth2CredentialSource implements CredentialSource {
+
+    /**
+     * Creates a new {@link Builder} instance in order to configure and build a {@link OAuth2CredentialSource}.
+     *
+     * @param tokenEndpointUrl the token endpoint that will be used to obtain OAuth2 access tokens
+     * @return a new builder instance
+     */
+    public static Builder builder(URL tokenEndpointUrl) {
+        return new Builder(tokenEndpointUrl);
+    }
+
+    private final String grantType;
+    private final URL tokenEndpointUri;
+    private final Consumer<Map<String, String>> consumer;
+    private String scopes;
+    private final Supplier<SSLContext> sslContextSupplier;
+    private final Supplier<HostnameVerifier> hostnameVerifierSupplier;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param grantType the OAuth2 grant type as defined by the specification
+     * @param tokenEndpointUri the OAuth2 Token Endpoint {@link URI}
+     * @param scopes a string with the scope of the access request
+     */
+    private OAuth2CredentialSource(String grantType, URL tokenEndpointUri, String scopes, Supplier<SSLContext> sslContextSupplier, Supplier<HostnameVerifier> hostnameVerifierSupplier) {
+        this(grantType, tokenEndpointUri, (Consumer<Map<String, String>>) stringStringMap -> {}, scopes, sslContextSupplier, hostnameVerifierSupplier);
```

Comment:
Better written like this:

```
this(grantType, tokenEndpointUri, (Map<String, String> map) -> {}, scopes, sslContextSupplier, hostnameVerifierSupplier);
```
Adding a type to the lambda argument is more readable than casting the lambda (that's something you should usually only do for lambdas that do not accept parameters but are accepted as arguments to overloaded methods).